soroban-env-host/src/test/budget_metering.rs

```diff
@@ -142,3 +144,51 @@ fn map_insert_key_vec_obj() -> Result<(), HostError> {
 
     Ok(())
 }
+
+#[test]
+fn test_recursive_type_clone() -> Result<(), HostError> {
+    let host = Host::test_host()
+        .test_budget(100000, 100000)
+        .enable_model(CostType::HostMemAlloc)
+        .enable_model(CostType::HostMemCpy);
+    let scmap: ScMap = host.map_err(
+        vec![
+            ScMapEntry {
+                key: ScVal::U32(1),
+                val: ScVal::U32(2),
+            },
+            ScMapEntry {
+                key: ScVal::U32(2),
+                val: ScVal::U32(4),
+            },
+        ]
+        .try_into(),
+    )?;
+    let v: Vec<Box<ScMap>> = vec![
+        Box::new(scmap.clone()),
+        Box::new(scmap.clone()),
+        Box::new(scmap.clone()),
+    ];
+
+    v.charge_for_clone(host.as_budget())?;
+    // Transition #                  (1)           (2)             (3)
+    // Type(size, count):  Vec(24,1) ---> Box(8,3) ----> Vec(24,3) ------> ScMapEntry(80,3x2)
+    // MemAlloc:                     8x3           24x3            80x3x2                     = 576
+    // MemCpy:             24        8x3  8x3      24x3  24x3      80x3x2  80x3x2             = 1176
+    // At the alloc boundary there is some double-charging happening.
+    // For example in transition #1, `Vec` charges for (1) allocation of 8x3 bytes, (2) copying of
+    // 8x3 bytes data into it. The `Box` charges for 8x3 for its own size.
```

Comment:
There is a bit of double counting going on. It comes from my current design and I’m not 100% sure if we can totally eliminate it: each type takes care itself (shallow copying of its own “pointer” size) and its substructures (allocating buffer, and shallow copying member “pointers” into the buffer). There isn’t really a way to know if itself is a substructure of something else, in which case its parent may have taken care of the shallow copying its pointer part. Thus leads to the shallow copy being double counted.